### PR TITLE
[Native][CPU][TopK] Improve perf by reducing swap operations

### PR DESCRIPTION
**Pull request by xziya**
Current implementation using `std::partial_sort` takes all pairs (keys and values) as its initial states which will lead to heavy memory footprint and access in some problems. This PR uses the first `k`-th data as the initial states in a `k`-length heap and go through all rest data with comparison to heap top and alternative heap push.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168

### CODE REVIEW

**Comment by janeyx99** [comment]
Sorry I only saw this right now--I'm likely not the right reviewer--but from a cursory glance, please include benchmarks!

**Comment by xziya** [comment]
> Sorry I only saw this right now--I'm likely not the right reviewer--but from a cursory glance, please include benchmarks!

Thanks for the kindly reminding. Here is the performance comparison on Apple M4 platform using the following script:

```python
import torch
import time

import numpy as np


warm_up = 10
dry_run = 50

def prb2str(shape, k):
    return f"{shape[0]}x{shape[1]}@{k}"

def bench(shape, k):
    torch.random.manual_seed(42)
    latencies = []
    src = torch.randn(shape)
    for _ in range(warm_up):
        dst_th, indices_th = torch.topk(
            src, k=k, dim=-1, sorted=True, largest=True)

    for _ in range(dry_run):
        _flush = torch.randn((1024, 1024))
        start = time.time()
        dst_th, indices_th = torch.topk(
            src, k=k, dim=-1, sorted=True, largest=True)
        end = time.time()
        latencies.append((end - start) * 1.e3)

    return np.mean(latencies)


if __name__ == "__main__":
    batch = [1, 64]
    hdim = [256, 1024, 4096]

    for b in batch:
        for h in hdim:
            shape = (b, h)
            for k_pow in range(int(np.log2(h)) + 1):
                latency = bench(shape, int(2 ** k_pow))
                print(f"{prb2str(shape, int(2 ** k_pow))}, {latency:.6f}")
```

The barplots of the latencies are

image
image
image

In the scenarios when `k * 64 < dim`, for example `1x4096@64`, this PR has a better latency than main branch. @janeyx99  @colesbury

File: aten/src/ATen/native/TopKImpl.h [hunk 1]
```diff
@@ -20,69 +20,124 @@ void topk_impl_loop(
...
+    const int64_t n) {
   // If k is zero, then output values and indices are empty tensors
   // So iterating over other dims is pointless
   if (k == 0) {
     return;
   }
   using elem_t = std::pair<accscalar_t, int64_t>;
-  std::vector<elem_t> queue(dim_size);
+
+  auto use_partial_sort = k * 64 <= dim_size;
```
Comment (r-barnes) [inline aten/src/ATen/native/TopKImpl.h]: Always use `const`
```suggestion
const auto use_partial_sort = k * 64 <= dim_size;
```

File: aten/src/ATen/native/TopKImpl.h [hunk 2]
```diff
@@ -20,69 +20,124 @@ void topk_impl_loop(
...
+      auto queue_begin = queue.begin();
+      auto queue_end = queue.begin() + k;
+      auto& queue_front_ref = queue.front();
+      auto& queue_back_ref = queue.back();
       if (largest) {
-        std::partial_sort(queue.begin(), queue.begin() + k, queue.end(),
-          [](const elem_t& x, const elem_t& y) -> bool {
-            return ((_isnan<accscalar_t>(x.first) && !_isnan<accscalar_t>(y.first)) || (x.first > y.first));
-          });
+        auto comp = [](elem_t const& a, elem_t const& b) {
```
Comment (r-barnes) [inline aten/src/ATen/native/TopKImpl.h]:
```suggestion
        const auto comp = [](elem_t const& a, elem_t const& b) {
```

File: aten/src/ATen/native/TopKImpl.h [hunk 3]
```diff
@@ -20,69 +20,124 @@ void topk_impl_loop(
...
+
+        if (sorted) {
+          std::sort_heap(queue_begin, queue_end, comp);
+        }
       } else {
-        std::partial_sort(queue.begin(), queue.begin() + k, queue.end(),
-          [](const elem_t& x, const elem_t& y) -> bool {
-            return ((!_isnan<accscalar_t>(x.first) && _isnan<accscalar_t>(y.first)) || (x.first < y.first));
-          });
+        auto comp = [](elem_t const& a, elem_t const& b) {
```
Comment (r-barnes) [inline aten/src/ATen/native/TopKImpl.h]: This whole section appears to be a copy of the previous section. Find a way to make the comparator the only thing that changes. Perhaps:
```c++
const auto comp = largest ? [](elem_t const& a, elem_t const& b) { ... } : [](elem_t const& a, elem_t const& b) { ... };
```

**Review by r-barnes — COMMENTED** [review]
Looks reasonable. I'm not sure how those small slow downs might affect things.

File: aten/src/ATen/native/TopKImpl.h [hunk 4]
```diff
@@ -20,69 +20,124 @@ void topk_impl_loop(
...
       }
     } else {
       if (largest) {
-        std::nth_element(queue.begin(), queue.begin() + k - 1, queue.end(),
-          [](const elem_t& x, const elem_t& y) -> bool {
-            return ((_isnan<accscalar_t>(x.first) && !_isnan<accscalar_t>(y.first)) || (x.first > y.first));
-          });
-        if (sorted) {
-          std::sort(queue.begin(), queue.begin() + k - 1,
+        std::nth_element(
```
Comment (r-barnes) [inline aten/src/ATen/native/TopKImpl.h]: Are all of these modifications only formatting changes? If so, it might be desirable to submit a separate, small PR in which we make the formatting changes and then make the substantiative changes in this PR - that'll make the change surface smaller if a revert is necessary.

File: aten/src/ATen/native/TopKImpl.h (hunk 3 above)
Comment (xziya) [reply aten/src/ATen/native/TopKImpl.h]: > This whole section appears to be a copy of the previous section. Find a way to make the comparator the only thing that changes. Perhaps:
> 
> ```c++
> const auto comp = largest ? [](elem_t const& a, elem_t const& b) { ... } : [](elem_t const& a, elem_t const& b) { ... };
> ```

Fixed the problem using a custom `TopKComparator` object and explicit template specification in each branch. Please take another review. Thanks.

**Comment by Skylion007** [comment]
> Looks reasonable. I'm not sure how those small slow downs might affect things.

topk is a pretty sensitive op for MOE routing according to Tri Dao's SonicMOE paper
